Migrate docs client webapp to React / Typescript / Material UI

anuraaga opened the pull request:
This large PR migrates the docs client from a Javascript/jQuery/Bootstrap-based architecture to Typescript/React/MaterialUI. 

It adds a separate `docs-client` project which is a standard NodeJS / webpack based build, run using `gradle-node-plugin`. Currently it's a separate JAR, but if it makes more sense to add it into the `core` JAR pretty sure we can do that too.

I changed the URLs a little, `/method` -> `/methods`, `/namedType` -> `/structs` + `/enums`. Since struct and enum pages are quite different, it made sense to separate the URLs out. The old URls are still supported and redirect to the new ones.

The TypeScript configuration is as strict as possible, including whitespace / semantic linting based on AirBnb style + Prettier. It's a pretty common configuration nowadays.

The `README` covers development concepts, let me know if there's anything else to add. 

To try out the client, I'd recommend pulling the branch and just stuffing a `Thread.sleep(Long.MAX_VALUE)` in one of `ThriftDocServiceTest` or `GrpcDocServiceTest`.

For #654

trustin commented on docs-client/3rd-party-licenses.txt:
```diff
@@ -0,0 +1,1618 @@
+THE FOLLOWING SETS FORTH ATTRIBUTION NOTICES FOR THIRD PARTY SOFTWARE THAT MAY BE CONTAINED IN PORTIONS OF THE @ARMERIA/DOCS CLIENT PRODUCT.
```
Question: Is this auto-generated? Would it be possible to have this generated into `/licenses`?

anuraaga replied:
The file is automatically generated by the command in the README. Do you mean move this file into `armeria/licenses/web-licenses.txt`? Can definitely do that.

trustin replied:
Yes, please. Could you also update `NOTICE.txt` to point at the `web-licenses.txt` there?

trustin commented on docs-client/src/components/VariableList/index.tsx:
```diff
@@ -0,0 +1,73 @@
+import Typography from '@material-ui/core/Typography';
+import React from 'react';
+
+import { Specification } from '../../lib/specification';
```
Global comment: The curly braces are redundant?

anuraaga replied:
I didn't add a "default export" to specification since it exports multiple things, and I've found in that case exporting a default can result in ugly looking imports like `import Specification, { Method, Service } from 'specification'`. I could add a default export though what do you think?

trustin replied:
I actually never knew `{ Something }` and `Something` are different. If they have different meaning, please leave them as they are.

trustin commented on docs-client/src/index.css:
```diff
@@ -0,0 +1,20 @@
+
+pre, code, kbd, var, samp, tt, .code {
+  font-family: 'Roboto Mono', 'Consolas', 'Menlo', 'Monaco', 'Lucida Console', 'Liberation Mono',
+    'DejaVu Sans Mono', monospace !important;
```
Could you avoid using `!important`?

anuraaga replied:
Since we're using material-ui components for the majority of our styling, and it defines it's own styles, it can be tricky to override fonts - the styles for each component are local to the component, so generally you would define stylized versions of each component you want to override. I thought it's probably overkill for this one single easy-to-grok CSS rule but can play with it some more if it's still a concern.

trustin replied:
Does this solution work? https://stackoverflow.com/a/45403932

anuraaga replied:
It applies the font to all typography, but this one is only for code snippets. By the way, the default font already uses `Roboto` per material-ui convention, so I didn't change the default typography.

trustin replied:
Ah I see. Let's leave this as it is for now.

trustin commented on docs-client/tsconfig.json:
```diff
@@ -0,0 +1,26 @@
+    "lib": ["esnext", "es2017", "es2016", "es2015", "dom"],
+    "module": "esnext",
+    "moduleResolution": "node",
+    "newLine": "LF",
```
Question: Does this determine the line separators of compilation output or the `.tsx` files?

anuraaga replied:
The compilation output, but since we only pipe though webpack anyways this doesn't matter so removed.

trustin replied:
Thanks!

trustin commented on settings.gradle:
```diff
@@ -7,6 +7,7 @@ includeWithFlags ':bom',                       'bom'
 
 // Published Java projects
 includeWithFlags ':core',                       'java', 'publish', 'shade', 'trim'
+includeWithFlags ':docs-client',                'publish'
```
Could we avoid publishing this but just use the build output into the resource set of `:core`?

anuraaga replied:
Yep both would work, though it would require a bit more configuration. I thought this might be a good idea to separate though since it's then very easy to exclude the jar to exclude the webapp if you don't care about it. What do you think?

trustin replied:
I'd still prefer making this part of the core JAR unless it adds too much. ;-)

trustin commented on gradle/scripts/lib/java-shade.gradle:
```diff
@@ -257,7 +257,7 @@ private Configuration createShadedTestRuntimeConfiguration(
         project.dependencies.add(shadedTestRuntime.name, files(recursedProject.tasks.trimShadedJar.outJarFiles))
-    } else {
+    } else if (recursedProject.tasks.findByName('shadedJar')) {
         project.dependencies.add(shadedTestRuntime.name, files(recursedProject.tasks.shadedJar.archivePath))
```
Thanks!

anuraaga replied:
I actually wasn't sure if I should edit this - should I edit the other gradle scripts repo instead?

trustin replied:
Let me send a PR for this to upstream.

trustin replied:
Fix has been pushed to `master`. Please rebase.

trustin commented:
@anuraaga Please rebase your PR so that Travis CI builds do not fail anymore due to JDK download failure.

trustin commented on core/build.gradle:
```diff
@@ -1,4 +1,6 @@
 dependencies {
+    compile project(':docs-client')
```
This has to be removed because otherwise the pom will refer to an unpublished artifact.

anuraaga replied:
Completely forgot to remove this! It's not needed since `builtBy` takes care of the task dependency.

trustin commented:
Miscellaneous issues found:
- There's no message when there are no parameters/exceptions/members, e.g. 'This method does not have parameters.'
- 'Use these HTTP headers for all functions.' checkbox does not work.

trustin commented:
From aesthetic PoV, I think it looks worse. :sweat_smile: But I guess we can improve on that later.

trustin commented:
![screenshot_2018-07-19 armeria documentation service](https://user-images.githubusercontent.com/173918/42920758-558c817e-8b53-11e8-9bce-a0aa24cbd9b9.png)

anuraaga commented:
Thanks - I found in the old code the only one that shows "There are no..." is the method parameters, the rest are just hidden so tried to reproduce that.

Also, sticky headers should be working if you submit the form. In the old code, it coincidentally sort of worked even if you didn't submit because state wasn't reinitialized, but if you went to a page with debug form to one without and back, they'd be gone. Is it better if the sticky kicks in before hitting submit button? Having three sources of the information (URL, JS state, specification) isn't terrible but a bit complicated.

And yeah happy to tweak aesthetic in this or followup PR (maybe a material design designer sits next to you? :D). In particular, I'm thinking of adding `Paper` in certain places, I think the grey background makes it look weird in many places.

trustin commented:
> Thanks - I found in the old code the only one that shows "There are no..." is the method parameters, the rest are just hidden so tried to reproduce that.

I see. It would be still nicer if 'there are no...' is shown for exceptions and fields as well.

> Also, sticky headers should be working if you submit the form. In the old code, it coincidentally sort of worked even if you didn't submit because state wasn't reinitialized, but if you went to a page with debug form to one without and back, they'd be gone. Is it better if the sticky kicks in before hitting submit button?

I think so. A user may have filled all the headers and then realized that he or she was on a wrong service method. :-)

> And yeah happy to tweak aesthetic in this or followup PR.

No worries. I'm not very sure what changes would make it look better yet anyway. Let me play with it a little bit once merged. :bowing_man:

anuraaga commented:
Thanks - think I addressed the comments. Also added some `Paper` so it's less grey and I'm sure we'll come up with more aesthetic tweaks in the future :)

trustin commented:
Thanks! It looks much better. :-)

Some more bugs:

- If you choose Enums -> `ConsistencyLevel` and scroll down the right pane, the left pane looks truncated.
- The left pane does not disappear into [hamburger](https://en.wikipedia.org/wiki/Hamburger_button) even if you shrink the width of the window.
- The HTTP headers text area should be visible if it's not empty. For example, if you tick the 'Use these HTTP headers for all functions' checkbox in `HelloService.hello()` and switch to `SleepService.sleep()`, the HTTP headers text area should be visible.

trustin commented:
As you see below, the left pane's height is not synchronized with the right pane's height.
![image](https://user-images.githubusercontent.com/173918/42935084-b6479ae6-8b83-11e8-8a4d-6441aa723d26.png)

anuraaga commented:
Thanks for the thorough check, had forgotten about the navigation. Now it's responsive, and also fixed the headers textbox so it shows when there are headers.

And after lots of playing, figured out the CSS magic to make the heights work I think!

trustin commented:
Great job! :-)

But, in mobile layout, after I click the hamburger, I find no way to hide the method list. I also see excessive top margin. Perhaps appbar should be there so a user can click the hamburger again to hide the method list?

anuraaga commented:
Ok render the mobile drawer at 80% width so clicking off of it can close it, this seems to be what e.g., Google Maps does. I also had the top-space just copied from the `material-ui.com` example, not sure why they had it seems fine without it :) Not just trying to figure out why their example renders sticky subheaders correctly but not mine, if I can't figure it out I'll just disable it.

anuraaga commented:
Got the method names to not overlap - think it's ready for another look

trustin commented:
More minor issues. :pray: 

Could we reduce the left margin of the hamburger button so it is aligned with the page content?

![image](https://user-images.githubusercontent.com/173918/42982420-35b71c46-8c1c-11e8-82a1-4f25945c406f.png)

Could we reduce the left margin of the left pane so that method names are placed better? Also, when a cursor hovers at a method name, the gray selection does not cover the row completely, leaving white background on the left of the method name.

Could we reduce the font size of the items in the left pane? It looks too big comparing to the right pane content. On the other hand, could the service names in the left pane be bigger?

I see a performance issue when pressing the hamburger button. It takes about a second since I clicked the button for the left pane to show up. Could we make it appear sooner?

anuraaga commented:
Fixed the method margin, it was introduced by extra `ul` tags that should have been `div`.

Tweaked the fonts so service name is larger and method names are smaller. Typography is a complex problem, and hoping to use the defaults as much as possible for now and customize it better in the future - WDYT?

And I'm not sure if I'm seeing the same performance issue, it feels like around 2-300ms for me, but I tried adding a prop that might make it faster for slower devices.

trustin commented on docs-client/src/containers/App/index.tsx:
```diff
@@ -102,26 +107,31 @@ function AppDrawer({ classes, navigateTo, specification }: AppDrawerProps) {
-                <ListSubheader className={classes.methodHeader}>
+            <div key={service.name}>
+              <ListSubheader className={classes.methodHeader}>
+                <Typography variant="title">
```
`subheading` looks better to me. What do you think?

anuraaga replied:
You're right, switched to it and also added `<code>`

anuraaga commented:
Thanks a lot for the detailed review!

anuraaga commented:
As a preview, looking forward to sending https://github.com/anuraaga/armeria/pull/5 with pluggable auth support after this is in :)

trustin commented:
@minwoox @hyangtack Gentle ping :stuck_out_tongue_winking_eye:

anuraaga commented on docs-client/src/containers/App/index.tsx:
```diff
@@ -0,0 +1,352 @@
+import Typography from '@material-ui/core/Typography';
+import MenuIcon from '@material-ui/icons/Menu';
+import React from 'react';
+import { hot } from 'react-hot-loader';
```
Yeah - the module name is on the right side. It is almost always all lowercase, but if it only exports a single class, sometimes it is uppercase (the case for the material ui exports). There is some history behind material ui exporting a class per module because webpack used to fail at tree shaking them when importing from the top-level. I feel this is probably not the case anymore, but they still document their imports in this form. In general it is considered bad practice for a library to not export everything from the top level though.

I stick to all lowercase all the time for modules for simplicity.

The left side is what is imported. A function is all lowercase while a class is camelcase.

Hope that answers the question :)

minwoox replied:
Ah! Didn't know that. Thank you for letting me know. :)

trustin commented:
Finally-! Thank you so much for your patience and contribution, @anuraaga !
